File: tests/device/hide/app-probe/app/src/main/java/dev/pathguard/hideprobe/StoragePath.java
package dev.pathguard.hideprobe;

import java.util.LinkedHashSet;

final class StoragePath {
    private static final String PRIMARY_ROOT = "/storage/emulated/0/";

    private StoragePath() {}

    static String relativeDirectory(String absolutePath) {
        if (!absolutePath.startsWith(PRIMARY_ROOT)
                || absolutePath.length() <= PRIMARY_ROOT.length()) {
            return null;
        }
        String relative = absolutePath.substring(PRIMARY_ROOT.length());
        while (relative.endsWith("/")) {
            relative = relative.substring(0, relative.length() - 1);
        }
        return relative.isEmpty() ? null : relative + "/";
    }

    static String[] expandVfsAliases(String[] canonicalPaths) {
        LinkedHashSet<String> paths = new LinkedHashSet<>();
        for (String canonicalPath : canonicalPaths) {
            paths.add(canonicalPath);
            String relative = relativeDirectory(canonicalPath);
            if (relative == null) continue;
            String withoutTrailingSlash = relative.substring(0, relative.length() - 1);
            paths.add("/sdcard/" + withoutTrailingSlash);
            paths.add("/storage/self/primary/" + withoutTrailingSlash);
            paths.add("/mnt/user/0/primary/" + withoutTrailingSlash);
            paths.add("/mnt/runtime/default/emulated/0/" + withoutTrailingSlash);
            paths.add("/mnt/runtime/read/emulated/0/" + withoutTrailingSlash);
            paths.add("/mnt/runtime/write/emulated/0/" + withoutTrailingSlash);
            paths.add("/data/media/0/" + withoutTrailingSlash);
        }
        return paths.toArray(new String[0]);
    }
}

File: tests/device/hide/app-probe/app/src/main/kotlin/dev/pathguard/hideprobe/StoragePath.kt
package dev.pathguard.hideprobe

internal object StoragePath {
    private const val PRIMARY_ROOT = "/storage/emulated/0/"

    fun relativeDirectory(absolutePath: String): String? {
        if (!absolutePath.startsWith(PRIMARY_ROOT) || absolutePath.length <= PRIMARY_ROOT.length) return null
        val relative = absolutePath.removePrefix(PRIMARY_ROOT).trimEnd { it == '/' }
        return relative.takeIf { it.isNotEmpty() }?.plus('/')
    }

    fun expandVfsAliases(canonicalPaths: Array<String>): Array<String> = buildList {
        canonicalPaths.forEach { path ->
            add(path)
            val relative = relativeDirectory(path)?.removeSuffix("/") ?: return@forEach
            add("/sdcard/$relative")
            add("/storage/self/primary/$relative")
            add("/mnt/user/0/primary/$relative")
            add("/mnt/runtime/default/emulated/0/$relative")
            add("/mnt/runtime/read/emulated/0/$relative")
            add("/mnt/runtime/write/emulated/0/$relative")
            add("/data/media/0/$relative")
        }
    }.distinct().toTypedArray()
}
